File: tests/testPsfAttributes.cc
// -*- LSST-C++ -*-

/* 
 * LSST Data Management System
 * Copyright 2008, 2009, 2010 LSST Corporation.
 * 
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the LSST License Statement and 
 * the GNU General Public License along with this program.  If not, 
 * see <http://www.lsstcorp.org/LegalNotices/>.
 */
 
//
// test a perfect Gaussian PSF and measure aperture photometry at different radii
//
#include <iostream>
#include <cmath>

#include "lsst/afw.h"
#include "lsst/meas/algorithms/SingleGaussianPsf.h"
#include "lsst/afw/geom/Angle.h"
#include "lsst/meas/algorithms/PSF.h"

#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE PsfAttributes

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunused-variable"
#include "boost/test/unit_test.hpp"
#pragma clang diagnostic pop
#include "boost/test/floating_point_comparison.hpp"

namespace measAlg = lsst::meas::algorithms;
namespace afwDetection = lsst::afw::detection;
namespace afwImage = lsst::afw::image;
namespace afwGeom = lsst::afw::geom;

BOOST_AUTO_TEST_CASE(PsfAttributes) {

    double sigma0 = 5.0;
    double aEff0 = 4.0*afwGeom::PI*sigma0*sigma0;
    
    int xwid = static_cast<int>(12*sigma0);
    int ywid = xwid;

    std::shared_ptr<afwDetection::Psf> psf(new measAlg::SingleGaussianPsf(xwid, ywid, sigma0));

    measAlg::PsfAttributes psfAttrib(psf, xwid/2.0, ywid/2.0);
    double sigma = psfAttrib.computeGaussianWidth(measAlg::PsfAttributes::ADAPTIVE_MOMENT);
    double m1    = psfAttrib.computeGaussianWidth(measAlg::PsfAttributes::FIRST_MOMENT);
    double m2    = psfAttrib.computeGaussianWidth(measAlg::PsfAttributes::SECOND_MOMENT);
    double noise = psfAttrib.computeGaussianWidth(measAlg::PsfAttributes::NOISE_EQUIVALENT);
    double bick  = psfAttrib.computeGaussianWidth(measAlg::PsfAttributes::BICKERTON);
    double aEff  = psfAttrib.computeEffectiveArea();
    
    std::cout << sigma0 << " " << sigma << std::endl;
    std::cout << sigma0 << " " << m1 << std::endl;
    std::cout << sigma0 << " " << m2 << std::endl;
    std::cout << sigma0 << " " << noise << std::endl;
    std::cout << sigma0 << " " << bick << std::endl;
    std::cout << aEff0 << " " << aEff << std::endl;
    
    BOOST_CHECK_CLOSE(sigma0, sigma, 1.0e-2);
    BOOST_CHECK_CLOSE(sigma0, m1, 3.0e-2);
    BOOST_CHECK_CLOSE(sigma0, m2, 1.0e-2);
    BOOST_CHECK_CLOSE(sigma0, noise, 1.0e-2);
    BOOST_CHECK_CLOSE(sigma0, bick, 1.0e-2);
    BOOST_CHECK_CLOSE(aEff0, aEff, 1.0e-2);

}

File: tests/test_psfAttributes.py
#!/usr/bin/env python
#
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
#
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the LSST License Statement and
# the GNU General Public License along with this program.  If not,
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""
Tests for PSF code

Run with:
   python testPsfAttributes.py
or
   python
   >>> import testPsfAttributes; psf.run()
"""
from __future__ import absolute_import, division, print_function
import math
import unittest

import lsst.afw.geom as afwGeom
import lsst.meas.algorithms as measAlg
import lsst.utils.tests

try:
    type(verbose)
    import lsst.afw.display.ds9 as ds9
except NameError:
    verbose = 0
    display = False

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-


class PsfAttributesTestCase(lsst.utils.tests.TestCase):

    def testGaussian(self):
        """Check that we can measure a single Gaussian's attributes."""
        sigma0 = 5.0
        aEff0 = 4.0*math.pi*sigma0**2

        xwid = int(12*sigma0)
        ywid = xwid

        # set the peak of the outer guassian to 0 so this is really a single gaussian.
        psf = measAlg.SingleGaussianPsf(xwid, ywid, sigma0)

        if False and display:
            im = psf.computeImage(afwGeom.PointD(xwid//2, ywid//2))
            ds9.mtv(im, title="N(%g) psf" % sigma0, frame=0)

        psfAttrib = measAlg.PsfAttributes(psf, xwid//2, ywid//2)
        sigma = psfAttrib.computeGaussianWidth(psfAttrib.ADAPTIVE_MOMENT)
        m1 = psfAttrib.computeGaussianWidth(psfAttrib.FIRST_MOMENT)
        m2 = psfAttrib.computeGaussianWidth(psfAttrib.SECOND_MOMENT)
        noise = psfAttrib.computeGaussianWidth(psfAttrib.NOISE_EQUIVALENT)
        bick = psfAttrib.computeGaussianWidth(psfAttrib.BICKERTON)
        aEff = psfAttrib.computeEffectiveArea()

        if verbose:
            print("Adaptive            %g v %g" % (sigma0, sigma))
            print("First moment        %g v %g" % (sigma0, m1))
            print("Second moment       %g v %g" % (sigma0, m2))
            print("Noise Equivalent    %g v %g" % (sigma0, sigma))
            print("Bickerton           %g v %g" % (sigma0, bick))
            print("Effective area      %g v %f" % (aEff0, aEff))

        self.assertLessEqual(abs(sigma0 - sigma), 1.0e-2)
        self.assertLessEqual(abs(sigma0 - m1), 3.0e-2)
        self.assertLessEqual(abs(sigma0 - m2), 1.0e-2)
        self.assertLessEqual(abs(sigma0 - noise), 1.0e-2)
        self.assertLessEqual(abs(sigma0 - bick), 1.0e-2)
        self.assertLessEqual(abs(aEff0 - aEff), 1.0e-2)

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-


class TestMemory(lsst.utils.tests.MemoryTestCase):
    pass


def setup_module(module):
    lsst.utils.tests.init()

if __name__ == "__main__":
    lsst.utils.tests.init()
    unittest.main()
